server/src-test/Hasura/Server/AuthSpec.hs

```diff
@@ -0,0 +1,408 @@
+                                , defaultRoleClaim .= ("user" :: Text)
+                                ]
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "editor")] mode
+              `shouldReturn` Right (mkRoleNameE "editor")
+            -- Uses the defaultRoleClaim:
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED")] mode
+              `shouldReturn` Right (mkRoleNameE "user")
+
+          it "rejects when requested role is not allowed" $ do
+            let claim = unObject [ allowedRolesClaim .= (["editor","user", "mod"] :: [Text])
+                                , defaultRoleClaim .= ("user" :: Text)
+                                ]
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "r00t")] mode
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "admin")] mode
+              `shouldReturn` Left AccessDenied
+
+          -- A corner case, but the behavior seems desirable:
+          it "always rejects when token has empty allowedRolesClaim" $ do
+            let claim = unObject [ allowedRolesClaim .= ([] :: [Text]), defaultRoleClaim .= ("user" :: Text) ]
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "admin")] mode
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "user")] mode
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED")] mode
+              `shouldReturn` Left AccessDenied
+
+          it "rejects when token doesn't have proper allowedRolesClaim and defaultRoleClaim" $ do
+            let claim0 = unObject [ allowedRolesClaim .= (["editor","user", "mod"] :: [Text]) ]
+                claim1 = unObject [ defaultRoleClaim .= ("user" :: Text) ]
+                claim2 = unObject []
+            for_ [claim0, claim1, claim2] $ \claim ->
+              getUserInfoWithExpTime claim [("Authorization", "IGNORED")] mode
+                `shouldReturn` Left JWTRoleClaimMissing
+
+    -- (*) FIXME NOTE (re above):
+    --
+    -- Ideally we should always return AccessDenied if the role we would
+    -- otherwise have returned does not match the requested role (from the
+    -- 'userRoleHeader').
+    --
+    -- This would harden a bit against bugs, makes the spec simpler, but
+    -- especially is better UX since in the current behavior the user can't be
+    -- sure which role their query is operating as (and in the worst case we
+    -- might e.g. delete rows the user didn't intend)
+    --
+    -- But this is a breaking change we need to think a little more about;
+    -- users might be relying on the behavior, e.g. just hardcoding a dev role
+    -- into clients.
```

Comment:
(I think my question of "default" versus "unauthenticated" role applies to this discussion as well: there are a couple of related terms here which would be nice to have clarified in a central place.)

Reply:
(also, again, just to check: do you want to solve this FIXME?)

Reply:
Added these references in comments:

-- [1]: https://hasura.io/docs/1.0/graphql/manual/auth/authentication/unauthenticated-access.html 
-- [2]: https://hasura.io/docs/1.0/graphql/manual/auth/authentication/jwt.html#the-spec

We have to punt on the FIXME for now. Requires more coordination. Vamshi is aware